docs/src/design/proposals/Registration.md

```diff
@@ -0,0 +1,518 @@
+  zero or more instances of objects of type `X` which have been registered,
+  using a `compositor` function provided as a constructor argument (or via
+  a subsequent `compose` call.)
+  * `register(factory : (function () : X), [options] : RegistrationOptions)`
+    registers an object (as returned by the provided `factory` function)
+    with this provider. Evaluation of the provided `factory` will be
+    deferred until the first `get()` call to the `Provider`.
+  * `compose(compositor : (function (X[]) : S), [options] : RegistrationOptions)`
+    introduces a new strategy for converting an array of registered objects
+    of type `X` into a single instance of an object of type `S`. The
+    highest-priority `compositor` that has been registered in this fashion
+    will be used to assemble the provided object (before decoration)
+* `ServiceProvider<S>` extends `CompositeProvider<S, S>` and provides analogous support
+  for the _composite services_
+  pattern used throughout Open MCT (which, in turn, is a superset of the
+  functionality needed for plain services.)
+* `Registry<T>` extends `CompositeProvider<T, T[]>` and provides analogous support for
+  _extension categories_, also
+  used ubiquitously through Open MCT.
+* `RegistrationOptions` provides a location for additional
+  registration-related metadata to be provided.
+  * `priority` is used to control the ordering and/or preference of
+    factories, compositors, and/or decorators,
+    [as in the existing Framework Layer.](http://nasa.github.io/openmctweb/guide/#priority)
+
+# Examples
+
+The following examples are provided to illustrate the intended usage of
+the Registration API. Particular attention is given to obeying and
+utilizing the "dependency injection as code style"
+[decision from the API Redesign](APIRedesign.md#decisions).
+
+## Building Applications on Open MCT
+
+Applications built using Open MCT are expected to extend the `MCT` base
+class and should typically self-install distinguishing plugins during
+the constructor call. Any pre-installed plugins that are undesirable
+should also be uninstalled at this point (to support such usage,
+`MCT` should expose instances of any installed plugins that are
+considered optional.)
+
+For example:
+
+```
+define(['mct', './plugins'], function (mct, plugins) {
+  var MCT = mct.MCT;
+
+  function Variant() {
+    MCT.call(this);
+
+    this.install(new plugins.SomePlugin());
+    this.install(new plugins.SomeOtherPlugin(this.core));
+
+    this.uninstall(this.plugins.persistence.localStorage);
+  }
+
+  Variant.prototype = Object.create(MCT.prototype);
+
+  return Variant;
+});
+```
+
+Running an application build using Open MCT then typically looks like:
+
+```
+define(['./Variant], function (Variant) {
+  new Variant().run();
+});
+```
+
+## Writing Plugins
+
+A plugin for Open MCT should inherit from the `Plugin` base class.
+
+Plugins will typically use extension points exposed by other plugins;
```

Comment:
This is an area where I disagree: Plugins should not expose extension points.  

The core framework should expose the full set of extension points, and plugins (batches of things that are registered to known extension points) may exist that provide default functionality for those extension points, but plugins should not expose extension points for use by other plugins.  (If they do, it is via their own means, and not via some platform sanctioned method).

Don't see a use case outside of the core platform for dependencies between plugins; all the current dependencies between plugins we have would be better expressed as "dependencies upon functionality of the core platform."  Or, they would be "customizations of a plugin".

WARP and VISTA themselves do not define new extension points-- they only reuse existing extension points.  I like that model.  It reduces the size of the API that people need to learn when they want to write their own plugin.

My concern is that we confuse the internal architecture of OpenMCT with the plugin architecture that is used to customize it for different purposes.  Yes, allowing plugins to define extension points means you can completely rewrite OpenMCT's functionality and it could be used for anything-- medical records, automated cat skinning, you name it.  But the fact is, we're OpenMCT.  We're not intended to be used for these things, we're intended to be used for mission control.